File: cpp/medium/ReorderList.cpp

#include <stack>
using namespace std;
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
};

void reorderList(ListNode *head) {
  ListNode *slow = head;
  ListNode *fast = head;
  while (fast->next) {
    fast = fast->next;
    if (!fast->next)
      break;
    fast = fast->next;
    slow = slow->next;
  }
  ListNode *another = slow->next;
  slow->next = nullptr;
  stack<ListNode *> s;
  while (another) {
    s.push(another);
    another = another->next;
  }
  slow = head;
  fast = head->next;
  while (!s.empty()) {
    ListNode *temp = s.top();
    s.pop();
    temp->next = fast;
    slow->next = temp;
    slow = fast;
    fast = fast->next;
  }
}

int main() {
  ListNode *root = new ListNode(1);
  root->next = new ListNode(2);
  root->next->next = new ListNode(3);
  root->next->next->next = new ListNode(4);
  reorderList(root);
  return 0;
}

File: go/middle/reorderList.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	// split list in tow equal length
	if head == nil || head.Next == nil {
		return
	}
	slow := head
	fast := head
	for fast.Next != nil && fast.Next.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}

	other := slow.Next
	slow.Next = nil
	// reverse other list
	var pre *ListNode = nil
	for other != nil {
		next := other.Next
		other.Next = pre
		pre = other
		other = next
	}

	for pre != nil {
		next := pre.Next
		pre.Next = head.Next
		head.Next = pre
		head = pre.Next
		pre = next
	}
}
